src/sims.cpp

```diff
@@ -1472,54 +1469,72 @@ namespace libsemigroups {
 
   }  // namespace sims
 
-  // TODO(0): (reiniscirpons) Change this in the same way as we do for Sims1,
-  // Once we add the citw stuff
-  SimsRefinerIdeals& SimsRefinerIdeals::init(Presentation<word_type> const& p) {
-    _presentation = p;
-    _knuth_bendices[0].init(congruence_kind::twosided, _presentation).run();
-    std::fill(
-        _knuth_bendices.begin() + 1, _knuth_bendices.end(), _knuth_bendices[0]);
+  ////////////////////////////////////////////////////////////////////////
+  // SimsRefinerIdeals
+  ////////////////////////////////////////////////////////////////////////
+
+  SimsRefinerIdeals& SimsRefinerIdeals::init() {
+    _presentation.init();
+    _knuth_bendices.clear();
+    _knuth_bendices.emplace(_default_thread_id, KnuthBendix_());
+
     return *this;
   }
 
-  bool SimsRefinerFaithful::operator()(Sims1::word_graph_type const& wg) {
-    auto first = _forbid.cbegin(), last = _forbid.cend();
-    // TODO(2) use 1 felsch tree per excluded pairs, and use it to check if
-    // paths containing newly added edges, lead to the same place
-    for (auto it = first; it != last; it += 2) {
-      bool this_rule_compatible = true;
-      for (uint32_t n = 0; n < wg.number_of_active_nodes(); ++n) {
-        auto l = word_graph::follow_path_no_checks(wg, n, *it);
-        if (l != UNDEFINED) {
-          auto r = word_graph::follow_path_no_checks(wg, n, *(it + 1));
-          if (r == UNDEFINED || (r != UNDEFINED && l != r)) {
-            this_rule_compatible = false;
-            break;
-          }
-        } else {
-          this_rule_compatible = false;
-          break;
-        }
-      }
-      if (this_rule_compatible) {
-        return false;
-      }
+  SimsRefinerIdeals&
+  SimsRefinerIdeals::operator=(SimsRefinerIdeals const& that) {
+    _default_thread_id = that._default_thread_id;
+    _knuth_bendices.clear();
+    _presentation = that._presentation;
+    // Don't copy _knuth_bendices because the thread id's will be wrong
+    _knuth_bendices.emplace(
+        _default_thread_id,
+        (*that._knuth_bendices.find(that._default_thread_id)).second);
+    return *this;
+  }
+
+  SimsRefinerIdeals& SimsRefinerIdeals::operator=(SimsRefinerIdeals&& that) {
+    _default_thread_id = std::move(that._default_thread_id);
+    _knuth_bendices    = std::move(that._knuth_bendices);
```

Comment:
In the copy assignment operator we explicitly don't copy the `_knuth_bendices` object since the thread ids will be wrong. Shouldn't we do the same here when move assigning? Since the thread ids will also presumably be wrong?

Reply:
I don't think we can actually call this operator when there are running threads, so I think this should be ok. The idea being that the lhs of `SimsRefinerIdeals x = std::move(y);` takes over the responsibility for the resources in `y`, so I think this is ok.